src/optim/FishLeg/fishleg.py

```diff
@@ -278,34 +283,99 @@ def update_aux(self) -> None:
+            nat_grads = []
+            p_idx = 0
+            for module in self.model.modules():
+                if not isinstance(module, nn.Sequential) and hasattr(module, "Qv"):
+                    layer_grads = []
+                    for name, param in module.named_parameters():
+                        if "fishleg_aux" not in name:
+                            layer_grads.append(param.grad.data)
+
+                    qv, qv_b = module.Qv(layer_grads)
+                    aux_loss -= torch.sum(grads[p_idx] * qv)
+                    aux_loss -= torch.sum(grads[p_idx + 1] * qv_b)
+
+                    aux_loss += damping / 2 * torch.sum(qv**2)
+                    aux_loss += damping / 2 * torch.sum(qv_b**2)
+
+                    nat_grads.append(qv)
+                    nat_grads.append(qv_b)
+                    p_idx += 2
+
+            # Add to gradients
+            p_idx = 0
+            for module in self.model.modules():
+                if not isinstance(module, nn.Sequential) and hasattr(module, "Qv"):
+                    layer_grads = []
+                    for name, param in module.named_parameters():
+                        if "fishleg_aux" not in name:
+                            param.data += self.eps * nat_grads[p_idx]
+                            p_idx += 1
+
+            pred_y = self.model(data_x)
+            plus_loss = self.likelihood.nll(pred_y, samples_y)
+
+            # Minus gradients
+            p_idx = 0
+            for module in self.model.modules():
+                if not isinstance(module, nn.Sequential) and hasattr(module, "Qv"):
+                    layer_grads = []
+                    for name, param in module.named_parameters():
+                        if "fishleg_aux" not in name:
+                            param.data -= 2 * self.eps * nat_grads[p_idx]
+                            p_idx += 1
+
+            pred_y = self.model(data_x)
+
+            minus_loss = self.likelihood.nll(pred_y, samples_y)
+
+            # Restore parameters
+            p_idx = 0
+            for module in self.model.modules():
+                if not isinstance(module, nn.Sequential) and hasattr(module, "Qv"):
+                    layer_grads = []
+                    for name, param in module.named_parameters():
+                        if "fishleg_aux" not in name:
+                            param.data += self.eps * nat_grads[p_idx]
+                            p_idx += 1
+
+            aux_loss += 0.5 * (plus_loss + minus_loss)
 
-                nat_grads = module.Qv(layer_grads)
+            aux_loss /= 1.0 + torch.norm(plus_loss.detach() + minus_loss.detach())
 
-                for n, (name, param) in enumerate(module.named_parameters()):
-                    if "fishleg_aux" not in name:
-                        nat_grad = nat_grads[n]
-                        sample_grad = grads[p_idx]
-                        true_grad = layer_grads[n]
-                        p_idx += 1
+        else:
```

Comment:
should be elif, with rank-1. and the else at the end should raise a not-implemented error